docs/source/format/CanonicalExtensions.rst

```diff
@@ -72,4 +72,30 @@ same rules as laid out above, and provide backwards compatibility guarantees.
 Official List
 =============
 
-No canonical extension types have been standardized yet.
+Fixed shape tensor
+==================
+
+* Extension name: `arrow.fixed_shape_tensor`.
+
+* The storage type of the extension: ``FixedSizeList`` where:
+
+  * **value_type** is the data type of individual tensors and
+    is an instance of ``pyarrow.DataType`` or ``pyarrow.Field``.
+  * **list_size** is the product of all the elements in tensor shape.
+
+* Extension type parameters:
+
+  * **value_type** = Arrow DataType of the tensor elements
+  * **shape** = shape of the contained tensors as a tuple
+  * **is_row_major** = boolean indicating the order of elements
```

Comment:
Can we not store `stride` instead? https://pytorch.org/docs/stable/generated/torch.Tensor.stride.html#torch.Tensor.stride

Reply:
In the [Zulip discussion](https://ursalabs.zulipchat.com/#narrow/stream/180245-dev/topic/Canonical.20extension.20type.20for.20tensors/near/324882917) we are leaning towards canonical type always storing row-major and letting applications store strides in metadata. Any arguments for or against from you or your users would be most welcome at this point!

Reply:
So you mean removing `is_row_major` as a parameter? I guess the issue boils down to how fast you want the loading/saving? In torch terminology, making a tensor contiguous (I'm guessing is the same as row major) makes a copy because the underlying memory representation changes. The same would apply to loading.

Reply:
> So you mean removing is_row_major as a parameter?

Yes. We would just use the physical layout of the source and not change memory layout when going in and out of the extension. We would provide an option to store the layout as metadata.

Reply:
> In torch terminology, making a tensor contiguous (I'm guessing is the same as row major) makes a copy because the underlying memory representation changes.

My understanding is that contiguous tensors in torch are indeed always row-major, but so that also means that if you have such a contiguous tensor, you don't need any copy to put this in the proposed extension TensorArray (or you can get it out without a copy).

Reply:
> Yes. We would just use the physical layout of the source and not change memory layout when going in and out of the extension. We would provide an option to store the layout as metadata.

This makes sense. I do think that in terms of UX it helps to be agnostic of those kind of consideration (and so having the format handle this in an underlying fashion it helps). But it makes sense to have a simple design.

> My understanding is that contiguous tensors in torch are indeed always row-major, but so that also means that if you have such a contiguous tensor, you don't need any copy to put this in the proposed extension TensorArray (or you can get it out without a copy).

Yes for contiguous it's perfect. The proposition of having `stride` as well, would make a zero copy system as well.

Reply:
Sometimes users will manipulate non_contiguous formats without realising. Typically `torch` uses a `to(memory_format=torch.channels_last)` which essentially is a equivalent to `.permute(0,2,3,1).contiguous()`. This API is often used in computer vision as some operation run faster when the tensors are not actually stored in "row major" fashion. So loading images in that specific format directly would help, which means storing non row major tensors.

```python
import torch

device = torch.device("cpu")
x = torch.randn(2,3,5,7, device=device)
y = x.to(memory_format=torch.channels_last)

# Assert that they don't share memory
do_share_memory = x.data_ptr() == y.data_ptr()
print(f"Channels Last shares memory: {do_share_memory}")

# Assert that channels last are not contiguous
print(f"Channels Last is contiguous: {y.is_contiguous()}")

# Data storage is equivalent to contiguous NHWC
z = x.permute(0,2,3,1).contiguous()
is_storage_equal = all(z_data == y_data for z_data, y_data in zip(z.storage(), y.storage()))
print(f"Channels Last storage is the same as the NHWC format: {is_storage_equal}")
```

More on this: https://pytorch.org/blog/accelerating-pytorch-vision-models-with-channels-last-on-cpu/

Reply:
Note that "contiguous" can be ambiguous. There are two possible meanings:
* physical contiguity: all values span a single contiguous area in memory (there are no "holes" without values in them), regardless of how they are ordered
* logical contiguity: values can be walked in either "C" or "F" order

Numpy uses the latter meaning. Note that logical contiguity implies physical contiguity, but the converse is not true, for example if you swap some dimensions in a more than 3+D array.

For now, it seems this spec and discussion is talking about _physical_ contiguity, as it doesn't care how the values are logically accessed (the interpretation of data is not part of the spec).

Reply:
Thus, by adding some metadata describing the dimension names in physical (row-major) order (for example `["N", "H", "W", "C"]`), we would leave the matter of logical contiguity to consumers of the data.

Reply:
(that said, there are situations where there are no dimension names, only dimension indices, such as when tensors are coming from Numpy arrays, so perhaps a slightly different convention is needed?)

Reply:
I think for this spec, the "physical" contiguity is a given, because that's the only possibility that it would match with our memory layout for FixedSizeList (or any other type, for that matter). 

But I do think we _also_ can say that the data should be interpreted as row-major ("logical" contiguity). At least, that's what we would use by default when converting to/from arrays (it's of course open for discussion if we should do that, but to be more explicit: that's what the current (example) implementations do), and applications could still ignore that and interpret the data however they want.

Reply:
> Thus, by adding some metadata describing the dimension names in physical (row-major) order (for example `["N", "H", "W", "C"]`), we would leave the matter of logical contiguity to consumers of the data.

But those dimension names only make sense if you _assume_ the data is in logical row-major order, no? Because otherwise if the data could be interpreted logically in different ways, you still don't know which logical dimension name matches with which actual data values?

Reply:
I see two problems with a "is_row_major" flag:
* when it's false, you don't know what the actual layout is (is it column-major, or is it in a non-trivial order like in the transpose() example)?
* saying that it "should be interpreted as row-major" doesn't say what the dimensions actually are. If you are shipping arrays Numpy then it's not a problem, because dimensions are anonymous (you are just doing abstract math on tensors). But if you are shipping ML tensors, the information might be incomplete (IIUC, you can have row-major "NHWC" or row-major "NCHW").

Reply:
> But those dimension names only make sense if you _assume_ the data is in logical row-major order, no?

The logical row-major order is given by the dimension names, so this is a bit circular?

Reply:
> (that said, there are situations where there are no dimension names, only dimension indices, such as when tensors are coming from Numpy arrays, so perhaps a slightly different convention is needed?)

So `dim_indexes` instead of `dim_names`?

Reply:
I think order of dimension names given by the application at creation time should define the logical contiguity and logical strides.

Reply:
Consequently a consumer can derive their own strides for a different logical view by (I think, didn't really check this):
```python
def strides(dim_names):
  dim_indexes = np.argsort(dim_names, physical_dim_names)
  view_shape = permute(shape, dim_indexes)
  return compute_row_major_strides(view_shape)
```

Reply:
> I see two problems with a "is_row_major" flag:

To be clear, I am not advocating for this flag. I am advocating for removing it, and state in the spec it is always assumed to be row major.

> saying that it "should be interpreted as row-major" doesn't say what the dimensions actually are

Yes, and for that reason we can have the optional dimension names

> The logical row-major order is given by the dimension names, so this is a bit circular?

I am not fully sure what's the consequence of it being circular, but does that then mean those two indeed go together? I.e. the order of the dimension names match the data if you assume it to be row-major (if that's the case, we still need to specify this, so you know how to match the list of dimension names to the data)

> So `dim_indexes` instead of `dim_names`?

I don't understand how that would be useful. That would just be [0, 1, 2] for 3D tensors? It's the optional name that could give meaning to an application (is it the width of the image? or the color channel?)

Reply:
> I don't understand how that would be useful. That would just be [0, 1, 2] for 3D tensors? It's the optional name that could give meaning to an application (is it the width of the image? or the color channel?)

I was referring to the numpy case where there are no `dim_names`. I do agree `dim_names` is the more useful option.

Reply:
> [@jorisvandenbossche ] @thomasw21 the page you link mentions the difference between "physical order" and "logical order". So the actual, physical layout in memory that pytorch uses can either be channels first (NCHW) or channels last (NHWC). But in both cases, this physical order is row-major. When interacting with those tensors as a user, pytorch however always shows the data as if it's NCHW, independent of the physical layout, i.e. what they call the "logical order". And it's when viewing the data as NCHW while the physical layout is NHWC that you get custom strides / non-C-contiguous layout for that view. But when considering the physical layout NHWC, the data is still row-major.

I think I have a hard time understanding this point. To me, torch has the ability to change the underlying "physical" ordering depending on how you interpret data. See Figure 1 of that blogpost
<img width="664" alt="image" src="https://user-images.githubusercontent.com/24695242/216576039-dfc056f5-b9cc-4c65-8ed1-d6ab9e2efe7d.png">

Reply:
I don't know if that helps the discussion:

```python
import torch

def get_1d_memory_buffer(tensor):
    return "".join(hex(elt) for elt in x.storage().untyped().byte())

x = torch.randn(2,3)
y = torch.empty(3,2).transpose(0,1)

# Fill y with x data
y[:] = x

assert x.shape == y.shape
assert get_1d_memory_buffer(x) != get_1d_memory_buffer(y)
# You can try printing `x` and `y` and you'll see that the tensors are the same
```

Reply:
> To me, torch has the ability to change the underlying "physical" ordering depending on how you interpret data. 

Yes, but both of those use row-major / C-contiguous memory layout. What you change is the order of the dimensions (C-H-W or H-W-C), and to do that while keeping row-major layout, changing between channels first/last layout requires actually shuffling the data in memory (and thus requires a copy). But either "physical order" is row-major and thus can be stored in the proposed FixedShapeTensor array without copy.

Reply:
> I don't know if that helps the discussion:

For your example, `x` and `y` indeed have a different layout of the values in memory, while seemingly representing the same tensor (in you print them). But that's because `y` is transposed after it was created (and this only changed the logical order, not the physical), and thus has custom strides. 

But you could still store both x and y without copy in a FixedShapeTensor array. The difference is that `x` would be stored as shape (2, 3), and y would be stored as (3, 2). And assume that in this dummy example the dimensions are called A and B, and if you want to keep track of the correct logical order, you would store the dimension names for `x` as `["A", "B"]` and for `y` as `["B", "A"]`. Then afterwards, after reading such data and the application knows it always wants the data in (A, B) order, it can transpose the data read from the stored `y` (just as you did when creating `y`)

Reply:
Btw I fixed a bug in my script ...

> The difference is that x would be stored as shape (2, 3), and y would be stored as (3, 2)

I understand that there exist a permutation of dimensions that allows me to get a "row major" format. I think it doesn't change how you store the permutation information, ie via dimension names or stride. It felt like a natural concept to me to store `stride`, as this would allow just provide a better generalisation IMO. But I do understand if the current extension would focus on pure row_major.

Reply:
OK, I understand (we might have been talking past each other a bit, as I was assuming you want to have `strides` to allow zero-copy for all cases, while I tried to convince you that it's not needed). 

It's certainly true that we _could_ store strides, but I am not sure it would be a better generalization of (or a full replacement for) dimension names. 
Consider for example that you have channels-last physical data (NHWC), but you view it as channels-first logically (NCHW). To store the data with the logical dimension order, this would requires a `strides` parameter. But assume you only store the strides in the `FixedShapeTensor` type and not the dimension names, then when consuming that data, you know the strides associated with it, but you still don't know for sure what the dimensions mean (because both NHWC viewed as NCHW, or NCHW viewed as NHWC would give you custom strides). Of course, if you know where the data is coming from and you know that it's from a pytorch context, then you can assume that the logical order is NCHW (that's how pytorch [always shows it](https://pytorch.org/blog/accelerating-pytorch-vision-models-with-channels-last-on-cpu/): "No matter what the physical order is, tensor shape and stride will always be depicted in the order of NCHW"), and that information combined with the strides ensures you know if the physical layout is channels-first or last. 
But that requires application-specific context to know that. While if you store the dimension names that match the order assuming a row-major layout, then you can infer the same information (and how to transpose it to get your desired logical order), but without requiring this application-specific knowledge (assuming different applications would use consistent dimension names, so you can recognize those).

So my current understanding is that dimension names are the more generalizable information. 

In addition, pushing the strides logic (how to translate the given dimension order and your desired dimension order to strides) to the application to deal with, keeps the implementation of the FixedShapeTensorType itself simpler, not requiring every implementation to deal with custom strides.